torch/csrc/jit/script/module.h

```diff
@@ -245,28 +186,28 @@ struct Module : public std::enable_shared_from_this<Module> {
     return methods.values();
   }
 
-
-  at::optional<NamedParameter&> find_parameter(const std::string& name) {
+  NamedParameter* find_parameter(const std::string& name) {
     return parameters.find(name);
   }
-  at::optional<NamedModule&> find_module(const std::string& name) {
+  NamedModule* find_module(const std::string& name) {
     return modules.find(name);
   }
-  at::optional<Method&> find_method(const std::string& name) {
-    if(auto pm = methods.find(name))
-      return at::optional<Method&>(**pm);
-    return at::nullopt;
+  Method* find_method(const std::string& name) {
+    if (auto method = methods.find(name)) {
+      return method->get();
+    }
+    return nullptr;
```

Comment:
Sooo the thing is, it's technically not valid to have an `optional<&T>` (it's usually not ever valid to store a reference in a container, and `optional` is a container). There is an explanation [here](https://stackoverflow.com/questions/26858034/stdoptional-specialization-for-reference-types), and it is also mentioned in [cppreference.com](http://en.cppreference.com/w/cpp/utility/optional):

> There are no optional references; a program is ill-formed if it instantiates an optional with a reference type. Alternatively, an optional of a std::reference_wrapper of type T may be used to hold a reference.

It worked here because we weren't using the `optional<T&>` in a way that breaks it, e.g.

```
int x = 5;
at::optional<int&> o;
o.emplace(x);
o = x; // should work according to API of std::optional, works for references, but compiler error!

e.cpp:90:5: error: use of deleted function ‘typename std::enable_if<(! std::is_same<typename std::decay<_Tp>::type, at::optional<T&> >::value), at::optional<T&>&>::type at::optional<T&>::operator=(U&&) [with U = int&; T = int; typename std::enable_if<(! std::is_same<typename std::decay<_Tp>::type, at::optional<T&> >::value), at::optional<T&>&>::type = at::optional<int&>&]’
   o = x;
     ^
```

The solutions are:
1. Continue illegal use of `at::optional` in this way, hoping we won't use it in a way that breaks
2. Use `at::optional<std::reference_wrapper<T>>`, but it makes accessing the value harder (`*o` doesn't give you a `T` directly ...)
3. Use the original type responsible for representing "a nullable reference to something": pointers, old school pointers

@zdevito (author of said code)

Reply:
Ok, let's use the pointer...